File: identity/src/main/java/com/android/identity/util/SimpleApplicationData.java
/*
 * Copyright 2023 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.android.identity.util;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.android.identity.internal.Util;

import java.io.ByteArrayInputStream;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.Executor;

import co.nstant.in.cbor.CborBuilder;
import co.nstant.in.cbor.CborDecoder;
import co.nstant.in.cbor.CborException;
import co.nstant.in.cbor.builder.MapBuilder;
import co.nstant.in.cbor.model.DataItem;
import co.nstant.in.cbor.model.SimpleValue;
import co.nstant.in.cbor.model.UnicodeString;

/**
 * An implementation of {@link ApplicationData} using a LinkedHashMap to back the key-value pairs.
 */
public class SimpleApplicationData implements ApplicationData {
    private Map<String, byte[]> mApplicationData = new LinkedHashMap<>();
    private final Listener mListener;

    /**
     * Creates a new SimpleApplicationData and sets the listener to be used for notification
     * when changes are made to the {@link SimpleApplicationData}.
     *
     * @param listener the listener or <code>null</code> if this functionality is not needed.
     */
    public SimpleApplicationData(@Nullable Listener listener) {
        mListener = listener;
    }

    @NonNull
    @Override
    public ApplicationData setData(@NonNull String key, @Nullable byte[] value) {
        if (value == null) {
            mApplicationData.remove(key);
        } else {
            mApplicationData.put(key, value);
        }

        if (mListener != null) {
            mListener.onDataSet();
        }
        return this;
    }

    @NonNull
    @Override
    public ApplicationData setString(@NonNull String key, @NonNull String value) {
        return this.setData(key, Util.cborEncodeString(value));
    }

    @NonNull
    @Override
    public ApplicationData setNumber(@NonNull String key, long value) {
        return this.setData(key, Util.cborEncodeNumber(value));
    }

    @NonNull
    @Override
    public ApplicationData setBoolean(@NonNull String key, boolean value) {
        return this.setData(key, Util.cborEncodeBoolean(value));
    }

    @Override
    public boolean keyExists(@NonNull String key) {
        return mApplicationData.get(key) != null;
    }

    @NonNull
    @Override
    public byte[] getData(@NonNull String key) {
        byte[] value = mApplicationData.get(key);
        if (value == null) {
            throw new IllegalArgumentException("This key is not present in the ApplicationData.");
        }
        return value;
    }

    @NonNull
    @Override
    public String getString(@NonNull String key) {
        byte[] value = this.getData(key);
        return Util.cborDecodeString(value);
    }

    @Override
    public long getNumber(@NonNull String key) {
        byte[] value = this.getData(key);
        return Util.cborDecodeLong(value);
    }

    @Override
    public boolean getBoolean(@NonNull String key) {
        byte[] value = this.getData(key);
        return Util.cborDecodeBoolean(value);
    }

    /**
     * Encode the {@link ApplicationData} as a byte[] using <a href="http://cbor.io/">CBOR</a>.
     *
     * @return a byte[] of the encoded app data.
     */
    public @NonNull byte[] encodeAsCbor() {
        CborBuilder appDataBuilder = new CborBuilder();
        MapBuilder<CborBuilder> appDataMapBuilder = appDataBuilder.addMap();
        for (String key : mApplicationData.keySet()) {
            appDataMapBuilder.put(key, mApplicationData.get(key));
        }
        appDataMapBuilder.end();
        return Util.cborEncode(appDataBuilder.build().get(0));
    }

    /**
     * Returns a fully populated SimpleApplicationData from a CBOR-encoded byte[] and sets the
     * listener to be used for notification when changes are made to the {@link SimpleApplicationData}.
     *
     * <p> To encode a SimpleApplicationData, use {@link #encodeAsCbor()}.
     *
     * @param encodedApplicationData The byte array resulting from {@link #encodeAsCbor()}.
     * @param listener the listener or <code>null</code> if this functionality is not needed.
     * @return A SimpleApplicationData with the correct key, value pairs.
     */
    public static @NonNull SimpleApplicationData decodeFromCbor(
            @NonNull byte[] encodedApplicationData, @Nullable Listener listener) {
        ByteArrayInputStream bais = new ByteArrayInputStream(encodedApplicationData);
        List<DataItem> dataItems;
        try {
            dataItems = new CborDecoder(bais).decode();
        } catch (CborException e) {
            throw new IllegalStateException("Error decoded CBOR", e);
        }
        if (dataItems.size() != 1) {
            throw new IllegalStateException("Expected 1 item, found " + dataItems.size());
        }
        if (!(dataItems.get(0) instanceof co.nstant.in.cbor.model.Map)) {
            throw new IllegalStateException("Item is not a map");
        }
        DataItem applicationDataDataItem = dataItems.get(0);

        SimpleApplicationData appData = new SimpleApplicationData(listener);
        for (DataItem keyItem : ((co.nstant.in.cbor.model.Map) applicationDataDataItem).getKeys()) {
            String key = ((UnicodeString) keyItem).getString();
            byte[] value = Util.cborMapExtractByteString(applicationDataDataItem, key);
            appData.mApplicationData.put(key, value);
        }
        return appData;
    }

    /**
     * Interface for listening for changes to the data in the {@link SimpleApplicationData}.
     *
     * <p>The {@link Listener#onDataSet()}  callback will be called every time the data stored
     * is modified (e.g. adding a key-value pair, changing a value, or removing a key-value pair).
     */
    public interface Listener {

        /**
         * Called when the data inside the {@link SimpleApplicationData} is changed
         */
        void onDataSet();
    }
}

File: identity/src/main/java/com/android/identity/util/SimpleApplicationData.kt
/*
 * Copyright 2023 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.identity.util

import co.nstant.`in`.cbor.CborBuilder
import co.nstant.`in`.cbor.CborDecoder
import co.nstant.`in`.cbor.CborException
import co.nstant.`in`.cbor.builder.MapBuilder
import co.nstant.`in`.cbor.model.DataItem
import co.nstant.`in`.cbor.model.Map
import co.nstant.`in`.cbor.model.UnicodeString
import com.android.identity.credential.NameSpacedData
import com.android.identity.credential.NameSpacedData.Companion.fromEncodedCbor
import com.android.identity.internal.Util
import com.android.identity.internal.Util.cborDecodeBoolean
import com.android.identity.internal.Util.cborDecodeLong
import com.android.identity.internal.Util.cborDecodeString
import com.android.identity.internal.Util.cborEncode
import com.android.identity.internal.Util.cborEncodeBoolean
import com.android.identity.internal.Util.cborEncodeNumber
import com.android.identity.internal.Util.cborEncodeString
import java.io.ByteArrayInputStream

/**
 * An implementation of [ApplicationData] using a [LinkedHashMap] as in-memory backing.
 *
 * Also contains functionality to serialize to/from CBOR.
 *
 * @param onDataChanged callback invoked whenever changes are made to a key that is
 * if it's added, changed, or removed.
 */
class SimpleApplicationData(private val onDataChanged: (key: String) -> Unit) : ApplicationData {

    private val data = mutableMapOf<String, ByteArray?>()

    override fun setData(key: String, value: ByteArray?): ApplicationData {
        if (value == null) {
            data.remove(key)
        } else {
            data[key] = value
        }
        onDataChanged(key)
        return this
    }

    override fun setString(key: String, value: String): ApplicationData =
        setData(key, cborEncodeString(value))

    override fun setNumber(key: String, value: Long): ApplicationData =
        setData(key, cborEncodeNumber(value))

    override fun setBoolean(key: String, value: Boolean): ApplicationData =
        setData(key, cborEncodeBoolean(value))

    override fun setNameSpacedData(key: String, value: NameSpacedData): ApplicationData =
        setData(key, value.encodeAsCbor())

    override fun keyExists(key: String): Boolean = data[key] != null

    override fun getData(key: String): ByteArray =
        data[key] ?: throw IllegalArgumentException("Key '$key' is not present")

    override fun getString(key: String): String = cborDecodeString(getData(key))

    override fun getNumber(key: String): Long = cborDecodeLong(getData(key))

    override fun getBoolean(key: String): Boolean = cborDecodeBoolean(getData(key))

    override fun getNameSpacedData(key: String): NameSpacedData = fromEncodedCbor(getData(key))

    /**
     * Encode the [ApplicationData] as a byte[] using [CBOR](http://cbor.io/).
     *
     * @return a byte[] of the encoded app data.
     */
    fun encodeAsCbor(): ByteArray {
        val appDataBuilder = CborBuilder()
        val appDataMapBuilder: MapBuilder<CborBuilder> = appDataBuilder.addMap()
        for (key in data.keys) {
            appDataMapBuilder.put(key, data[key])
        }
        appDataMapBuilder.end()
        return cborEncode(appDataBuilder.build().get(0))
    }

    companion object {
        /**
         * Returns a fully populated SimpleApplicationData from a CBOR-encoded
         * [ByteArray] and sets the listener to be used for notification when
         * changes are made to the [SimpleApplicationData].
         *
         * To encode a SimpleApplicationData, use [.encodeAsCbor].
         *
         * @param encodedApplicationData The byte array resulting from [.encodeAsCbor].
         * @param onDataChanged callback invoked whenever changes are made to a key that is
         * if it's added, changed, or removed.
         * @return A [SimpleApplicationData].
         */
        @JvmStatic
        fun decodeFromCbor(
            encodedApplicationData: ByteArray,
            onDataChanged: (key: String) -> Unit
        ): SimpleApplicationData {
            val bais = ByteArrayInputStream(encodedApplicationData)
            val dataItems: List<DataItem> = try {
                CborDecoder(bais).decode()
            } catch (e: CborException) {
                throw IllegalStateException("Error decoding CBOR", e)
            }
            check(dataItems.size == 1) { "Expected 1 item, found " + dataItems.size }
            check(dataItems[0] is Map) { "Item is not a map" }
            val applicationDataDataItem: DataItem = dataItems[0]
            val appData = SimpleApplicationData(onDataChanged)
            for (keyItem in (applicationDataDataItem as Map).keys) {
                val key: String = (keyItem as UnicodeString).string
                val value = Util.cborMapExtractByteString(applicationDataDataItem, key)
                appData.data[key] = value
            }
            return appData
        }
    }
}